rust/arrow/src/datatypes.rs

```diff
@@ -1296,9 +1301,31 @@ impl Field {
             nullable,
             dict_id,
             dict_is_ordered,
+            metadata: None,
         }
     }
 
+    /// Sets the `Field`'s optional custom metadata.
+    /// The metadata is set as `None` for empty map.
+    /// Return cloned `Self`.
+    #[inline]
+    pub fn with_metadata(&mut self, metadata: Option<BTreeMap<String, String>>) -> Self {
```

Comment:
I think we shouldn't pass a mutable self, then return a new self, because this is going to mutate the incoming `Field` and return a clone of it, i.e.:

```rust
let mut field = Field::new("my_field", DataType::Int32, false);
let field2 = field.with_metadata(Some(metadata));
assert_eq!(field, field2); // this will be true, so we're creating 2 fields
```

It would be better to return nothing from the function, as we've already mutated `&mut self`.

Reply:
@nevi-me I totally agree that the cloned behavior is a bug instead of coding style problem.
I had fixed it per your suggestion and renamed the function from `with_metadata` to `set_metadata` without any return value.

Reply:
Thanks, I'm happy with the rename